plugins/hls-change-type-signature-plugin/src/Ide/Plugin/ChangeTypeSignature.hs

```diff
@@ -0,0 +1,237 @@
+                                                           , _isPreferred = Nothing
+                                                           , _disabled    = Nothing
+                                                           , _edit        = Just $ mkChangeSigEdit uri declSrcSpan (mkNewSignature declName actualType)
+                                                           , _command     = Nothing
+                                                           , _xdata       = Nothing
+                                                           }
+
+mkChangeSigTitle :: Text -> Text -> Text
+mkChangeSigTitle declName actualType = "change signature for ‘" <> declName <> "’ to: " <> actualType
+
+mkChangeSigEdit :: Uri -> RealSrcSpan -> Text -> WorkspaceEdit
+mkChangeSigEdit uri ss replacement =
+        let txtEdit = TextEdit (realSrcSpanToRange ss) replacement
+            changes = Just $ Map.singleton uri (List [txtEdit])
+        in WorkspaceEdit changes Nothing Nothing
+
+mkNewSignature :: Text -> Text -> Text
+mkNewSignature declName actualType = declName <> " :: " <> tidyActualType actualType
+
+-- | Cleans up type signatures with oddly-named type variables (a0, m0 ...)
+tidyActualType :: Text -> Text
+tidyActualType actualType =
+    replaceAllFTyVars
+    $ replaceAllMTyVars
+    $ replaceAllTTyVars
+    $ replaceAllUniqueTyVars uniqueTyVars actualType
+    where
+        findAllTyVars :: Text -> [Text]
+        findAllTyVars regex = getAllTextMatches (actualType =~ regex)  :: [Text]
+        uniqueTyVars = nub $ findAllTyVars "[abcdeghijklnopqrsuvwxyz][0-9]"
+        -- specialize searching on any `f`, `m` or `t` tyVars
+        fTyVars = findAllTyVars "f[0-9]"
+        mTyVars = findAllTyVars "m[0-9]"
+        tTyVars = findAllTyVars "t[0-9]"
+        -- | From a list of type variables replace all instances with the list of freshVars provided
+        replaceAllTyVars :: [Text] -> Text -> [Text] -> Text
+        replaceAllTyVars freshVars sig = snd . foldl replaceWithFreshVar (freshVars, sig)
+        replaceAllUniqueTyVars replace sig = replaceAllTyVars freshUniqueVars sig replace
+        replaceAllFTyVars sig = replaceAllTyVars (freshVars "f") sig fTyVars
+        replaceAllMTyVars sig = replaceAllTyVars (freshVars "m") sig mTyVars
+        replaceAllTTyVars sig = replaceAllTyVars (freshVars "t") sig tTyVars
+
+
+freshVars :: Text -> [Text]
+freshVars var = var : [var <> T.pack (show i) | i <- [0..]]
+
+freshUniqueVars :: [Text]
+freshUniqueVars = map T.singleton validVars <> [ a <> b | a <- freshUniqueVars, b <- freshUniqueVars ]
+    where
+        validVars = filter (`notElem` ['f', 'm', 't']) ['a'..'z']
+
+
+replaceWithFreshVar :: ([Text], Text) -> Text -> ([Text], Text)
+replaceWithFreshVar (replacement:rest, haystack) needle = (rest, T.replace needle replacement haystack)
+replaceWithFreshVar _ _ = error "How do I handle this..., the pattern match for infinite list shouldn't fail"
```

Comment:
The replacement list is always infinite, is there a way to encode this so I don't need to have the error match case?